pkg/credentials/credentials.go

```diff
@@ -283,13 +287,45 @@ func defaultValidationFunc(creds map[string]string) error {
 }
 
 func openstackValidationFunc(creds map[string]string) error {
-	for k, v := range creds {
-		if k == OpenStackTenantID || k == OpenStackTenantName {
-			continue
+	alwaysRequired := []string{OpenStackAuthURL, OpenStackDomainName, OpenStackRegionName}
+
+	for _, key := range alwaysRequired {
+		if v, ok := creds[key]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required but is not present", key)
 		}
-		if len(v) == 0 {
-			return errors.Errorf("key %v is required but isn't present", k)
+	}
+
+	var (
+		appCredsInUse  bool
+		userCredsInUse bool
+	)
+
+	if v, ok := creds[OpenStackApplicationCredentialID]; ok && len(v) != 0 {
+		if v, ok := creds[OpenStackApplicationCredentialSecret]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required if %v is present", OpenStackApplicationCredentialSecret, OpenStackApplicationCredentialID)
 		}
+		appCredsInUse = true
+	}
+
+	if v, ok := creds[OpenStackUserName]; !appCredsInUse && ok && len(v) != 0 {
+		if v, ok := creds[OpenStackPassword]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required but is not present", OpenStackPassword)
+		}
+		userCredsInUse = true
+	}
```

Comment:
Similar to the other comment, what if the password is provided but not the username?

Also, I'm not sure about the `!appCredsInUse` check -- if the app credentials are used, `userCredsInUse` will always be false (because of `!appCredsInUse`), so the validation on L324 (`appCredsInUse && userCredsInUse`) doesn't have any effect. If this is correct, we should probably drop `!appCredsInUse`.

Reply:
Should be addressed now.